File: 3° semestre/Tecnicas-de-programacão-2/aula-24-08-16/AbstractFactory/CodigoDoProfessor/Professor_AbstractFactory01.py
# Interface da Fabrica Abstrata:
class FabricaAbstrata:
    def criaProdutoA(self):
        pass
    def criaProdutoB(self):
        pass
    def criaProdutoC(self):
        pass

# Fabrica Concreta 1 - Cria Produtos do Tipo A, B e C:
class FabricaConcreta1(FabricaAbstrata):
    def criaProdutoA(self):
        return ProdutoConcretoA1()
    
    def criaProdutoB(self):
        return ProdutoConcretoB1()
    
    def criaProdutoC(self):
        return ProdutoConcretoC1()

# Fabrica Concreta 2 - Cria Produtos do Tipo A, B e C:
class FabricaConcreta2(FabricaAbstrata):
    def criaProdutoA(self):
        return ProdutoConcretoA2()
    
    def criaProdutoB(self):
        return ProdutoConcretoB2()
    
    def criaProdutoC(self):
        return ProdutoConcretoC2()

# Interface dos Produtos do Tipo A ---------------------------------------------:
class ProdutoAbstratoA:
    def metodoA(self):
        pass

# Implementação Concreta do Produto do Tipo A - Fabrica 1:
class ProdutoConcretoA1(ProdutoAbstratoA):
    def metodoA(self):
        return "Produto do Tipo A da Fabrica 1"

# Implementação Concreta do Produto do Tipo A - Fabrica 2:
class ProdutoConcretoA2(ProdutoAbstratoA):
    def metodoA(self):
        return "Produto do Tipo A da Fabrica 2"

# Interface dos Produtos do Tipo B ---------------------------------------------:
class ProdutoAbstratoB:
    def metodoB(self):
        pass

# Implementação Concreta do Produto do Tipo B - Fabrica 1:
class ProdutoConcretoB1(ProdutoAbstratoB):
    def metodoB(self):
        return "Produto do Tipo B da Fabrica 1"

# Implementação Concreta do Produto do Tipo B - Fabrica 2:
class ProdutoConcretoB2(ProdutoAbstratoB):
    def metodoB(self):
        return "Produto do Tipo B da Fabrica 2"

# Interface dos Produtos do Tipo C ---------------------------------------------:
class ProdutoAbstratoC:
    def metodoC(self):
        pass

# Implementação Concreta do Produto do Tipo C - Fabrica 1:
class ProdutoConcretoC1(ProdutoAbstratoC):
    def metodoC(self):
        return "Produto do Tipo C da Fabrica 1"

# Implementação Concreta do Produto do Tipo C - Fabrica 2:
class ProdutoConcretoC2(ProdutoAbstratoC):
    def metodoC(self):
        return "Produto do Tipo C da Fabrica 2"

# Função para Demonstrar o Padrão Abstract Factory-------------------------------:
def codigoCliente(factory):
    produtoA = factory.criaProdutoA()
    produtoB = factory.criaProdutoB()
    produtoC = factory.criaProdutoC()

    print(produtoA.metodoA())
    print(produtoB.metodoB())
    print(produtoC.metodoC())

# Exemplo de Uso da 1ª Fabrica:
factory1 = FabricaConcreta1()
codigoCliente(factory1)

# Exemplo de Uso da 2º Fabrica:
factory2 = FabricaConcreta2()
codigoCliente(factory2)

File: 3° semestre/Tecnicas-de-programacão-2/aula-24-08-16/AbstractFactory/CodigoDoProfessor/Professor_AbstractFactory01.js
// Interface da Fabrica Abstrata:
class FabricaAbstrata{
    criaProdutoA(){}
    criaProdutoB(){}
    criaProdutoC(){}
}

// Fabrica Concreta 1 - Cria Produtos do Tipo A, B e C:
class FabricaConcreta1 extends FabricaAbstrata{
    criaProdutoA(){
        return new ProdutoConcretoA1();
    }

    criaProdutoB(){
        return new ProdutoConcretoB1();
    }

    criaProdutoC(){
        return new ProdutoConcretoC1();
    }
}

// Fabrica Concreta 2 - Cria Produtos do Tipo A, B e C:
class FabricaConcreta2 extends FabricaAbstrata{
    criaProdutoA(){
        return new ProdutoConcretoA2();
    }

    criaProdutoB(){
        return new ProdutoConcretoB2();
    }

    criaProdutoC(){
        return new ProdutoConcretoC2();
    }
}

// Interface dos Produtos do Tipo A ---------------------------------------------:
class ProdutoAbstratoA{
    metodoA(){}
}

// Implementação Concreta do Produto do Tipo A - Fabrica 1:
class ProdutoConcretoA1 extends ProdutoAbstratoA{
    metodoA(){
        return "Produto do Tipo A da Fabrica 1";
    }
}

// Implementação Concreta do Produto do Tipo A - Fabrica 2:
class ProdutoConcretoA2 extends ProdutoAbstratoA{
    metodoA(){
        return "Produto do Tipo A da Fabrica 2";
    }
}

// Interface dos Produtos do Tipo B ---------------------------------------------:
class ProdutoAbstratoB{
    metodoB(){}
}

// Implementação Concreta do Produto do Tipo B - Fabrica 1:
class ProdutoConcretoB1 extends ProdutoAbstratoB{
    metodoB(){
        return "Produto do Tipo B da Fabrica 1";
    }
}

// Implementação Concreta do Produto do Tipo B - Fabrica 2:
class ProdutoConcretoB2 extends ProdutoAbstratoB{
    metodoB(){
        return "Produto do Tipo B da Fabrica 2";
    }
}

// Interface dos Produtos do Tipo C ---------------------------------------------:
class ProdutoAbstratoC{
    metodoC(){}
}

// Implementação Concreta do Produto do Tipo C - Fabrica 1:
class ProdutoConcretoC1 extends ProdutoAbstratoC{
    metodoC(){
        return "Produto do Tipo C da Fabrica 1";
    }
}

// Implementação Concreta do Produto do Tipo C - Fabrica 2:
class ProdutoConcretoC2 extends ProdutoAbstratoC{
    metodoC(){
        return "Produto do Tipo C da Fabrica 2";
    }
}

// Função para Demonstrar o Padrão Abstract Factory-------------------------------:
function codigoCliente(factory){
    const produtoA = factory.criaProdutoA();
    const produtoB = factory.criaProdutoB();
    const produtoC = factory.criaProdutoC();

    console.log(produtoA.metodoA());
    console.log(produtoB.metodoB());
    console.log(produtoC.metodoC());
}

// Exemplo de Uso da 1ª Fabrica:
const factory1 = new FabricaConcreta1();
codigoCliente(factory1);

// Exemplo de Uso da 2º Fabrica:
const factory2 = new FabricaConcreta2();
codigoCliente(factory2);
